### PR TITLE
[zeromq] Cherry pick patches to fix build issues

### LINKED ISSUE #21623: ZeroMQ BUILD_FAILED
I'm attempting to install zeromq:x64-windows (fails with x86 too), and it results with BUILD_FAILED for 4.3.4#2. 

package: zeromq[core]:x64-windows -> 4.3.4#2
vcpkg version: 30465138e 2021-11-22 (10 hours ago)
vcpkg-tool version: 2021-11-15-ec76affcc987efbf172b7da0b05df711b48fa535

[config-x64-windows-out.log](https://github.com/microsoft/vcpkg/files/7588847/config-x64-windows-out.log)
[install-x64-windows-dbg-out.log](https://github.com/microsoft/vcpkg/files/7588848/install-x64-windows-dbg-out.log)
[patch-x64-windows-0-err.log](https://github.com/microsoft/vcpkg/files/7588849/patch-x64-windows-0-err.log)
[stdout-x64-windows.log](https://github.com/microsoft/vcpkg/files/7588850/stdout-x64-windows.log)
[x64-windows.vcpkg_abi_info.txt](https://github.com/microsoft/vcpkg/files/7588851/x64-windows.vcpkg_abi_info.txt)

### PR DESCRIPTION
**Pull request by LilyWangLL**
Fixes #21623
Update to latest commit, the upstream latest release version is 4.3.4 which commit on 2021-01-17. 
All features test pass on `x86-windows`.

### CODE REVIEW

**Review by JackBoosY — APPROVED** [review]

**Comment by ras0219-msft** [comment]
Can we cherry pick https://github.com/zeromq/libzmq/pull/4310 instead, since `libzmq` has an official numbered release scheme?

**Comment by LilyWangLL** [comment]
> Can we cherry pick [zeromq/libzmq#4310](https://github.com/zeromq/libzmq/pull/4310) instead, since `libzmq` has an official numbered release scheme?

Thanks for your suggestion. I have updated commit to https://github.com/zeromq/libzmq/pull/4311, it is a secondary repair of https://github.com/zeromq/libzmq/pull/4310.

**Review by JackBoosY — APPROVED** [review]

File: versions/z-/zeromq.json [hunk 1]
```diff
@@ -1,5 +1,15 @@
...
+    {
+      "git-tree": "60790ade9af3dd66e1e4029650a60e050c4206c9",
+      "version-date": "2021-11-25",
+      "port-version": 0
+    },
+    {
+      "git-tree": "eb07369f7e7cac53a1450f36e25f8c01ac33cc5f",
+      "version-date": "2022-01-19",
+      "port-version": 0
+    },
```
Comment (JackBoosY) [inline versions/z-/zeromq.json:12]:
```suggestion
```

**Comment by ras0219-msft** [comment]
Can we just download the diff files and apply them onto the released version instead of pulling the commit?

Specifically, `vcpkg_download_distfile()` on

https://patch-diff.githubusercontent.com/raw/zeromq/libzmq/pull/4310.diff
https://patch-diff.githubusercontent.com/raw/zeromq/libzmq/pull/4311.diff

and add them to `PATCHES` of the main source extraction.

**Comment by LilyWangLL** [comment]
> Can we just download the diff files and apply them onto the released version instead of pulling the commit?
> 
> Specifically, `vcpkg_download_distfile()` on
> 
> https://patch-diff.githubusercontent.com/raw/zeromq/libzmq/pull/4310.diff https://patch-diff.githubusercontent.com/raw/zeromq/libzmq/pull/4311.diff
> 
> and add them to `PATCHES` of the main source extraction.

Done.

File: ports/zeromq/portfile.cmake [hunk 2]
```diff
@@ -1,9 +1,24 @@
+vcpkg_download_distfile(patch4310
+    URLS "https://patch-diff.githubusercontent.com/raw/zeromq/libzmq/pull/4310.diff"
+    FILENAME "4310.diff"
```
Comment (ras0219-msft) [inline ports/zeromq/portfile.cmake:3]: These filenames are very likely to conflict with other ports; can we instead name them like `zeromq-libzmq-4310.diff`?

**Review by ras0219-msft — CHANGES_REQUESTED** [review]

**Review by JackBoosY — APPROVED** [review]

**Comment by zmughal** [comment]
This approach is causing build issues for me.

If you run:
```shell
curl https://patch-diff.githubusercontent.com/raw/zeromq/libzmq/pull/4311.diff | sha512sum
```
repeatedly, the result can be either of the following:

```
2b04e0ce4743d27070ea832c45e2d8fa0091c755757937cfa2a2bb43283ee38dc9f27343989e1ad8c45fda8a3cfaa012250b0c581e2f0407938cbb61b2a21e63  4311.diff
08c3f000eb1582c94473d83c05b64f8934a14d18446faa10a396b8425eba108363cbb3382b7c95ffea8589ed5312244cd3cf76c3e49db20071dfae7c5d9e94ed  4311.diff
```

This is because the line for `index` from git can differ by one character:
```diff
+ index 24cce83d11..c6d625188d 100644
- index 24cce83d1..c6d625188 100644
```

**Comment by JackBoosY** [comment]
@zmughal The hash always the second one when I repeate run the command you provided.

**Comment by zmughal** [comment]
@JackBoosY, not sure exactly why it is different each time for me (different servers?), but the second one is different from the hash listed in this PR. Which is why I have been seeing build failures.

**Comment by JackBoosY** [comment]
@zmughal Related stackoverflow: https://stackoverflow.com/questions/54188267/how-to-control-hash-length-in-git-diff-header

I should check the github document.

**Comment by zmughal** [comment]
I have submitted feedback to GitHub here: <https://github.com/github/feedback/discussions/12531>.

**Pull request merged**